### ISSUE TITLE
GOPATH environment variable not current until restart of vscode

### ISSUE THREAD

**Issue by krum110487** [Microsoft/vscode-go#546]
Here is my situation, the environment variables used for vscode commands are based on when vscode is ran.

If I want to change my project and change the folder in vscode without closing vscode, I will get an error as it is checking for a dependency that doesn't exist in the old path. Of course it works just fine if I run the build command from a terminal, just not from within vscode when it saves.

would it be possible to have a `source ~/.bashrc` command run before each save logic to make sure we have the most recent variables? Is this something that is possible? Otherwise if there is a way I can change the path in vscode from outside vscode so it is updated with my custom bash script, that would work as well.

ps. I am using ubuntu 16.04 LTS

**Comment by ramya-rao-a** [comment]
@krum110487 GOPATH supports multiple workspaces. Will that not help?
Correct me if I am wrong, I am guessing you have the below situation

GOPATH: `/usr/me/mygopath`
First project: `/usr/me/mygopath/src/github.com/me-again/my-go-project1`
Second project: `/usr/you/some-folder/src/github.com/me-again/my-go-project2`

You have the first project open in VS Code and everything is fine. You open the folder for the second project and get errors that it cannot find packages.

**Comment by krum110487** [comment]
so I have something like this:
`/usr/me/projects` is my folder that holds all of my projects

but as of now I have a "project" folder inside to differentiate src folders.

so I have `/usr/me/projects/projectGroup/src/site/project/go-project1`
The idea was for me to keep all of my go projects separate as I don't want just 1 single path. I have read it can be a pain for a lot of projects to have the same path (although this is something that seems pretty heavily debated when it is brought up)

What I do is I set my gopath to `/usr/me/Projects/projectName` then open visual studio code, and everything works fine. Then without closing visual studio I update my gopath to the other project and change the folder. Of course vscode isn't picking up the new variables as it is using the ones it started with (which I understand why, just curious if there is a work around)

Worst case scenario I can get along fine with restarting vscode, I am just not sure if this is desired functionality or I am simply using go incorrectly.

**Comment by ramya-rao-a** [comment]
@krum110487 Thanks for clarifying!

The `go.gopath` setting does override the $GOPATH set in the terminal from when VSCode was originally launched. I just tried it. 

- Open the first project, everything is fine
- Open the second project, observe errors (red squiggly lines on the imports) as dependent packages are not found
- Update `go.gopath` in the settings to point to the second project
- Errors in the go file do not go away
- Make some change in the go file and save so that build/vet/linting gets triggered
- All is good

Can you confirm?

**Comment by krum110487** [comment]
Hmmm, this is strange, I cannot seem to replicate my error, I change my $GOPATH to `\usr\me\Projects` which is a folder that should not work, Then I removed the go.path settings and restarted VSCode and there are no errors with my imports, it even shows the correct location.

This is a bit baffling to me, I have no idea where it is getting the location from, is there a way to determine how the gopath is being resolved, from everything I know it should not be possible. I will try playing around with it to see if I can break it, because it seems like it should not be working, but it is throwing no errors right now. (but it should be)

**Comment by ramya-rao-a** [comment]
The .vscode folder stores your workspace settings. Deleting that folder deletes all your workspace settings.
For your situation setting `go.gopath` in the workspace settings for each of your projects should work if you do not want to have colon separated multiple workspaces in your GOPATH

gopath from env is overriden by gopath in user setting which is overridden by go path in workspace

**Comment by krum110487** [comment]
Yea, I get how it is supposed to work, but I removed the go.gopath setting and it still wasn't erroring. It didn't error until I removed the .vscode folder and all the settings inside. I will see if I can reproduce the issue as it is acting a bit funky for me.

**Comment by ramya-rao-a** [comment]
Remember that between each change you make to the `go.gopath` in the settings, you need to re-trigger the build (make a change -> save) to see the errors appear/disappear.

**Comment by krum110487** [comment]
Correct, I did do that, I am working on seeing if I can re-create the weirdness that was occurring.

**Comment by krum110487** [comment]
Here are the steps to reproduce the funky things I have seen:

1. Set the $GOPATH to an invalid location that should error
2. Open

**Comment by ramya-rao-a** [comment]
At Step 7, I see the error
At step 9, I see the error
At step 11,12,14 I see the error

I am not able to repro :(

Which version of the Go extension are you using?

Can you try opening the output window and keep an eye on the "Go" channel there? That is where the build output is shown and any errors as well. 
Can you check if you are seeing build output there with/without errors when you expect errors?
Or you dont see any build output at all

**Comment by krum110487** [comment]
I believe I am using the latest, I will check when I get home today.

I had the output window with the GO channel on the entire time, it was showing up and not having errors.
I am testing this specifically with an invalid import, so maybe that changes things?

as a background, I used apt-get to install vscode, then upgraded it in the program, then installed go plugin and upgraded it. I am running Ubuntu 16.04 LTS.

I installed the tools manually to my own folder and added that folder to my PATH so they can be accessed, instead of having them in my direct path on my workspace.
